swim/suspicion.go

```diff
@@ -33,62 +33,85 @@ type suspect interface {
 	incarnation() int64
 }
 
+type timer struct {
+	*time.Timer
+	// at which state the timer was started
+	state string
+}
+
 // Suspicion handles the suspicion sub-protocol of the SWIM protocol
 type suspicion struct {
 	sync.Mutex
 
 	node *Node
 
-	timeout time.Duration
-	timers  map[string]*time.Timer
-	enabled bool
-	logger  log.Logger
+	suspectTimeout time.Duration
+	timers         map[string]*timer
+	enabled        bool
+	logger         log.Logger
 }
 
 // newSuspicion returns a new suspicion SWIM sub-protocol with the given timeout
-func newSuspicion(n *Node, timeout time.Duration) *suspicion {
+func newSuspicion(n *Node, suspectTimeout time.Duration) *suspicion {
 	suspicion := &suspicion{
-		node:    n,
-		timeout: timeout,
-		timers:  make(map[string]*time.Timer),
-		enabled: true,
-		logger:  logging.Logger("suspicion").WithField("local", n.Address()),
+		node:           n,
+		suspectTimeout: suspectTimeout,
+		timers:         make(map[string]*timer),
+		enabled:        true,
+		logger:         logging.Logger("suspicion").WithField("local", n.Address()),
 	}
 
 	return suspicion
 }
 
-func (s *suspicion) Start(suspect suspect) {
-	s.withLock(func() {
-		if !s.enabled {
-			s.logger.Warn("cannot start suspect period while disabled")
-			return
-		}
+func (s *suspicion) StartSuspect(suspect suspect) {
```

Comment:
The type `suspect` is also out of place here since it is more than the suspect, it is more of a `subject` of the change.